### ISSUE TITLE
Custom Gson, singleton mode, deserialization in multiple coroutines, resulting in crash

### ISSUE THREAD

**Issue by mnlin0905** [google/gson#2698]
# Gson version
Gson: `2.8.9`

# Java / Android version
Android `33`
Java `11`
Kotlin `1.7.20`

# Used tools
gradle `7.2`

AndroidStudio:
```
Android Studio Hedgehog | 2023.1.1 Patch 2
Build #AI-231.9392.1.2311.11330709, built on January 19, 2024
Runtime version: 17.0.7+0-b2043.56-10550314 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 16
Registry:
    external.system.auto.import.disabled=true
    debugger.watches.in.variables=false
    debugger.new.tool.window.layout=true
    debugger.mayBringFrameToFrontOnBreakpoint=false
    ide.text.editor.with.preview.show.floating.toolbar=false
    ide.instant.shutdown=false
    ide.experimental.ui=true
    ide.balloon.shadow.size=0
    ide.images.show.chessboard=true

Non-Bundled Plugins:
    cn.haojiyou.CodeGlance3 (2.1.1-2021.1.3)
    com.intellij.ideolog (222.3.2.0)
    carbon-now-sh (1.3.0)
    String Manipulation (9.12.0)
    lermitage.intellij.extra.icons (2024.3.4)
    DBN (3.4.3211.0)
    com.github.copilot (1.5.2.5345)
    izhangzhihao.rainbow.brackets (2023.3.7)
    cn.yiiguxing.plugin.translate (3.5.7)
```

no ProGuard  (debug)

# Description

### When using (exceptions) through singleton examples

```kotlin
Private fun generateGsonForOUTS(): Gson{
Return GsonBuilder()
. registerTypeAdapter (CommonAction:: class. java, InjectActionForCommonAction)
. create()
}

val outsGson = generateGsonForOUTS()
```

```kotlin
Private fun generateGsonForOUTS(): Gson{
Return GsonBuilder()
. registerTypeAdapter (CommonAction:: class. java, InjectActionForCommonAction)
. create()
}
```

```kotlin
withContext(Dispatchers.Default) {
it.nodes.flatMap { group ->
group.childNodes.map { node ->
async {
withContext(Dispatchers.IO) {
innerReadRootView(node)
}
}
}
}.awaitAll()
}
```

There will be exceptions .....

when i change

```
val outsGson = generateGsonForOUTS()
```

 to :

```kotlin
val outsGson: Gson
      get() = generateGsonForOUTS()
```

No more abnormalities will occur !

## Expected behavior

Want to use singleton mode normally

## Actual behavior

When using singleton mode in this scenario, it can cause a crash

# Reproduction steps

NA

# Exception stack trace

```log
W  com.google.gson.JsonSyntaxException: java.lang.IllegalStateException
W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:226)
W  	at com.google.gson.Gson.fromJson(Gson.java:963)
W  	at com.google.gson.Gson.fromJson(Gson.java:928)
D  applyConfigurationToAppResourcesLocked app.getDisplayId() return callback.displayId:-1
W  	at com.google.gson.Gson.fromJson(Gson.java:877)
W  	at com.google.gson.Gson.fromJson(Gson.java:848)
W  	at com.oplus.demo.naturalcomm.outssupport.OUTSRepo.readOUTSFromFile$lambda$12$innerReadRootView(OUTSRepo.kt:137)
W  	at com.oplus.demo.naturalcomm.outssupport.OUTSRepo.access$readOUTSFromFile$lambda$12$innerReadRootView(OUTSRepo.kt:23)
W  	at com.oplus.demo.naturalcomm.outssupport.OUTSRepo$readOUTSFromFile$2$1$1$1$1$1.invokeSuspend(OUTSRepo.kt:151)
W  	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
W  	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
W  	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
W  	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
W  	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
W  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
W  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
W  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
W  Caused by: java.lang.IllegalStateException
W  	at com.google.gson.Gson$FutureTypeAdapter.read(Gson.java:1049)
W  	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
W  	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
W  	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
W  	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
W  	... 15 more
```

**Comment by Marcono1234** [comment]
This sounds like #625, which was fixed in Gson 2.10.1. Can you please try that version, or the latest version (currently 2.11.0)?

**Comment by mnlin0905** [comment]
> This sounds like #625, which was fixed in Gson 2.10.1. Can you please try that version, or the latest version (currently 2.11.0)?

It should have been resolved. After upgrading and testing for a few days, it can work normally. Thank you
